internal/pkg/deploy/upload/asset/asset.go

```diff
@@ -0,0 +1,117 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+// Package customresource provides functionality to upload Copilot custom resources.
+package asset
+
+import (
+	"fmt"
+	"io"
+	"path/filepath"
+
+	"github.com/spf13/afero"
+)
+
+type filterAction int
+
+const (
+	include filterAction = iota + 1
+	exclude
+)
+
+// UploadFunc is the function signature to upload contents to a destination.
+type UploadFunc func(dest string, contents io.Reader) (url string, err error)
+
+// UploadInput is the input of Upload.
+type UploadInput struct {
+	Source      string
+	Destination string
+	Includes    []string
+	Excludes    []string
+	Upload      UploadFunc
+	Reader      *afero.Afero
+}
+
+// Upload uploads static assets to Cloud Storage.
+func Upload(in *UploadInput) ([]string, error) {
+	files, err := listFiles(in.Reader, in.Source)
+	if err != nil {
+		return nil, err
+	}
+	filter := filter{buildPatterns(in.Includes, in.Excludes)}
+	filteredFiles, err := filter.apply(files)
```

Comment:
lol this whole package is so similar to my interview question for SDE2s 😂 

What do you think of a layout like this overall:
```go
type filepathMatcher interface {
   match(path string) (bool, error)
}

type includeMatcher string
func (f includeMatcher) match(path string) (bool, error)

type excludeMatcher string
func (f excludeMatcher) match(path string) (bool, error)

type andMatcher []filepathMatcher
func (f andMatcher) match(path string) (bool, error)
```

and in `Upload` we get something like:
```go
includeMatchers := buildincludeMatchers(opts.Include...)
excludeMatchers := buildExcludeMatchers(opts.Exclude...)
matcher := buildAndMatcher(append(includeMatchers, excludeMatchers...))

var fpaths []string
err := afero.Walk(fs, source, walkFnWithMatcher(fpaths, matcher))
```